server/commands/list_objects.go

```diff
@@ -0,0 +1,232 @@
+
+	timeoutCtx := ctx
+	var cancel context.CancelFunc
+	if q.listObjectsDeadlineInSeconds != 0 {
+		timeoutCtx, cancel = context.WithTimeout(ctx, time.Second*time.Duration(q.listObjectsDeadlineInSeconds))
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, req.StoreId, req.AuthorizationModelId, req.Type, req.Relation, req.User, req.ContextualTuples, resultsChan, errChan, resolvedChan)
+	}()
+
+	select {
+	case <-timeoutCtx.Done():
+		return nil, serverErrors.NewInternalError("Timeout exceeded", timeoutCtx.Err())
+	case <-resolvedChan:
+		objectIDs := make([]string, 0)
+		for objectID := range resultsChan {
+			objectIDs = append(objectIDs, objectID)
+		}
+
+		return &openfgapb.ListObjectsResponse{
+			ObjectIds: objectIDs,
+		}, nil
+	case genericError := <-errChan:
+		return nil, genericError
+	}
+}
+
+func (q *ListObjectsQuery) ExecuteStreamed(ctx context.Context, req *openfgapb.StreamedListObjectsRequest, srv openfgapb.OpenFGAService_StreamedListObjectsServer) error {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return err
+	}
+
+	resultsChan := make(chan string, q.listObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	var cancel context.CancelFunc
+	if q.listObjectsDeadlineInSeconds != 0 {
+		timeoutCtx, cancel = context.WithTimeout(ctx, time.Second*time.Duration(q.listObjectsDeadlineInSeconds))
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, req.StoreId, req.AuthorizationModelId, req.Type, req.Relation, req.User, req.ContextualTuples, resultsChan, errChan, resolvedChan)
+	}()
+
+	for {
+		select {
+		case <-timeoutCtx.Done():
+			return serverErrors.NewInternalError("Timeout exceeded", timeoutCtx.Err())
+		case objectID, ok := <-resultsChan:
+			if ok {
+				err := srv.Send(&openfgapb.StreamedListObjectsResponse{
+					ObjectId: objectID,
+				})
+				if err != nil {
+					return serverErrors.NewInternalError("", err)
+				}
+			} else {
+				// the channel was closed
+				return nil
+			}
+		case genericError, ok := <-errChan:
+			if ok {
+				return serverErrors.NewInternalError("", genericError)
+			}
+		}
+	}
+}
+
+func (q *ListObjectsQuery) getUniqueObjects(ctx context.Context, storeID, targetObjectType string, ctxTuples *openfgapb.ContextualTupleKeys) ([]string, error) {
+	uniqueObjects, err := q.datastore.ReadUniqueObjects(ctx, storage.ReadRelationshipTuplesFilter{
+		StoreID:            storeID,
+		OptionalObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+
+	if ctxTuples != nil {
+		// Include the objects in the contextual tuples in the result
+		uniqueSet := make(map[string]bool, len(uniqueObjects))
+		for _, o := range uniqueObjects {
+			_, found := uniqueSet[o]
+			if !found {
+				uniqueSet[o] = true
+			}
+		}
+
+		for _, ctxTuple := range ctxTuples.TupleKeys {
+			_, found := uniqueSet[ctxTuple.Object]
```

Comment:
I'm guessing it matters the `type` if you are using `targetObjectType` otherwise it's just noise 🤔

Reply:
Excellent point!